contracts/src/TeleporterTokenSource.sol

```diff
@@ -235,75 +266,91 @@ abstract contract TeleporterTokenSource is
                 destinationAddress: input.destinationBridgeAddress,
                 feeInfo: TeleporterFeeInfo({
                     feeTokenAddress: input.primaryFeeTokenAddress,
-                    amount: feeAmount
+                    amount: input.primaryFee
                 }),
                 requiredGasLimit: input.requiredGasLimit,
                 allowedRelayerAddresses: new address[](0),
                 message: abi.encode(message)
             })
         );
 
-        if (isMultiHop) {
-            emit TokensRouted(messageID, input, adjustedAmount);
-        } else {
-            emit TokensSent(messageID, _msgSender(), input, adjustedAmount);
-        }
+        emit TokensRouted(messageID, input, adjustedAmount);
     }
 
     function _sendAndCall(
         bytes32 sourceBlockchainID,
         address originBridgeAddress,
         address originSenderAddress,
         SendAndCallInput memory input,
-        uint256 amount,
-        bool isMultiHop
+        uint256 amount
     ) internal sendNonReentrant {
+        _validateSendAndCallInput(input);
+
+        // Require that a single hop transfer does not have a multi-hop fallback recipient.
         require(
-            input.recipientContract != address(0),
-            "TeleporterTokenSource: zero recipient contract address"
-        );
-        require(input.requiredGasLimit > 0, "TeleporterTokenSource: zero required gas limit");
-        require(input.recipientGasLimit > 0, "TeleporterTokenSource: zero recipient gas limit");
-        require(
-            input.recipientGasLimit < input.requiredGasLimit,
-            "TeleporterTokenSource: invalid recipient gas limit"
-        );
-        require(
-            input.fallbackRecipient != address(0),
-            "TeleporterTokenSource: zero fallback recipient address"
+            input.multiHopFallback == address(0),
+            "TeleporterTokenSource: non-zero multi-hop fallback"
         );
```

Comment:
Same here re: adding a dedicated validation helper to match the destination contract.